server/pulp/server/lazy/url.py

```diff
@@ -0,0 +1,594 @@
+        for k, v in parsed.items():
+            p = '{k}={v}'.format(k=k, v=quote(v))
+            encoded.append(p)
+        return ';'.join(encoded)
+
+
+class Key(object):
+    """
+    Provides RSA key management.
+    """
+
+    @staticmethod
+    def load(path=None, pem=None):
+        """
+        Get/Load an RSA key at the specified path.
+
+        :param path: An absolute path to a PEM encoded key.
+        :type path: str
+        :param pem: A PEM encoded key.
+        :type pem: str
+        :return: The loaded key.
+        :rtype: RSA.RSA
+        """
+        if path:
+            with open(path) as fp:
+                pem = fp.read()
+        bfr = BIO.MemoryBuffer(pem)
+        if 'PRIVATE' in pem:
+            key = RSA.load_key_bio(bfr)
+        else:
+            key = RSA.load_pub_key_bio(bfr)
+        return key
+
+
+class URL(object):
+    """
+    An URL object that supports signing.
+
+    :ivar content: The actual URL.
+    :type content: ParseResult
+    """
+
+    POLICY = 'policy'
+    SIGNATURE = 'signature'
+
+    def __init__(self, content):
+        """
+        :param content: The *actual* URL key.
+        :type content: str
+        """
+        self.content = urlparse(content)
+
+    @property
+    def scheme(self):
+        """
+        :return: The *scheme* component of the URL.
+        :rtype: str
+        """
+        return self.content.scheme
+
+    @property
+    def netloc(self):
+        """
+        :return: The *host:port* component of the URL.
+        :rtype: str
+        """
+        return self.content.netloc
+
+    @property
+    def path(self):
+        """
+        :return: The *path* component of the URL.
+        :rtype: str
+        """
+        return self.content.path
+
+    @property
+    def params(self):
+        """
+        :return: The *params* component of the URL.
+        :rtype: str
+        """
+        return self.content.params
+
+    @property
+    def query(self):
+        """
+        :return: The *query* component of the URL.
+        :rtype: str
+        """
+        return self.content.query
+
+    @property
+    def resource(self):
+        """
+        :return: The *resource* specified by the URL.
+        :rtype: str
+        """
+        path = self.path
+        params = self.params
+        query = self.query
+        if params:
+            params = ';{p}'.format(p=params)
+        if query:
+            query = '?{q}'.format(q=query)
+        resource = [
+            path,
+            params,
+            query
+        ]
+        return ''.join(resource)
+
+    def sign(self, key, expiration=90, **extensions):
+        """
+        Sign the URL using the specified private RSA key.
+        Has the format of: <url>?policy=<policy>;signature=<signature>.
+        The *policy* is: {resource: <resource>, expiration: <seconds>, extensions: <ext>}
+        The *resource* is the path?query in the original URL.
+        The *expiration* is: seconds since epoch.
```

Comment:
This says 'seconds since epoch' but the expiration is just 90 :hourglass_flowing_sand: . Is it seconds since creation?